### DISCUSSION TITLE
How can I curve the vertices with SimplePlane?

### DISCUSSION THREAD

**Discussion by gapgag55** [pixijs/pixijs#8263] [Q&A] [answered]
### Expected Behavior

I am trying to use SimplePlane to change the mesh of the image as Photoshop does.

I face the problem of how to curve the line between points as an example,

![image](https://user-images.githubusercontent.com/4579539/161701856-7f355bde-df1f-41fd-99fb-4cc92136d591.png)

### Current Behavior

The actual one is

![image](https://user-images.githubusercontent.com/4579539/161701819-20bee230-03e7-44a2-8f83-75ec912a4a4c.png)

To complete the goal, I need to change the image as Photoshop does which is curve the mesh 

```
    const app = new PIXI.Application({
      width: 1024,
      height: 1024,
      view: document.getElementById('canvas'),
      backgroundColor: '0x000000',
    });

    const container = new PIXI.Container()
    container.x =  52
    container.y =  189
    app.stage.addChild(container)

    const plane = new PIXI.SimplePlane(new PIXI.Texture.from('./images/mug.png'), 4, 4)
    plane.x = 0
    plane.y = 0
    plane._bounds.minX = 0
    plane._bounds.minY = 0
    plane._bounds.maxX = 919
    plane._bounds.maxY = 645
    container.addChild(plane)
    
    const buffer = plane.geometry.getBuffer('aVertexPosition');
    buffer.data[2] = 305.7120241581259
    buffer.data[3] = 33.09457357247439 
```

### Environment
pixi.js version: master and latest release 6.3.0
Nuxt.js version: 3.0.0

**Comment by miltoncandelero** [comment]
What pixi does is an affine transformation and what you most likely want is a perspective correct transformation

![image](https://user-images.githubusercontent.com/15253010/161751353-36fa16f4-8ae4-4cb6-ab04-d68045a4cc0b.png)

You need to write your own shader, you can find how to do it here:

https://www.reedbeta.com/blog/quadrilateral-interpolation-part-1/

---

OR you can use pixi-projection:
https://pixijs.io/examples/#/plugin-projection/quad-homo.js
https://pixijs.io/examples/#/plugin-projection/quad-bi.js

you can find it here: https://github.com/pixijs/pixi-projection

**Comment by gapgag55** [comment]
Thanks for replying @miltoncandelero 

I want to know more. Can Pixi projection curve the line as the image below? As I know, the projection can do only 4 points affine transformation. 

PS. I changed the image by using Wrap perspective in Photoshop

![image](https://user-images.githubusercontent.com/4579539/161753561-17efac07-a5de-47f9-a7af-3934dcfa7775.png)

**Comment by miltoncandelero** [comment]
I think it should be able to do something like that 🤔 

![image](https://user-images.githubusercontent.com/15253010/161760098-869c37ea-3893-4f06-80e0-fae265e981b3.png)
https://pixijs.io/examples/#/plugin-projection/quad-bi.js

**Reply by gapgag55** [reply]
The point is that, how can I do more than 4 points. As my image above, it has 16 points based on the wrap grid from Photoshop.

**Comment by bigtimebuddy** [comment]
The short answer is we don't have anything in Pixi that tries to replicate the warp feature in Photoshop, so you'll need to build this yourself. 

Here's how I might approach this problem: there's a UI (9-grid with bezier handles) and a warped image. Creating the UI controls could be done using Graphics and should be straightforward. You could use that information (location of the bezier handles) to drive a SimplePlane with a much higher resolution mesh than you were using initially and weight the position of SimplePlane's mesh points depending on the position of the bezier handles from the UI. 

There also might be a shader approach like Milton suggested.

**Reply by gapgag55** [reply]
Do you have an example code for the approach one? I also use that actually, but the result does not curve and I need it to shape smoothly like SimpleRope does. 

![image](https://user-images.githubusercontent.com/4579539/161808838-55b6a74a-0307-480b-8b12-c910162a7478.png)

**Reply by bigtimebuddy** [reply]
Sorry, I do not have an example. I'm describing a general approach you might take.

**Comment by gapgag55** [comment, marked as answer]
Solved, Just use Point and Grid bezier from this Codepen.

https://codepen.io/osublake/pen/EmNMXY/8785174690eabaf5ba99f785b9d3b64e?editors=0011

**Reply by ivanpopelyshev** [reply]
Thanks for the example, I'll save it. Its old stuff, and pixi-projection with a bit more code might allow bezier curves , that way it will require less geometry. Unfortunately, like i said, before you, feature wasnt required. With your material it might be done later , so thank you!
